// blooket/src/assets/links.js
module.exports = {
    links: {
        join: "https://api.blooket.com/api/firebase/join",
        joinQuery: "https://api.blooket.com/api/firebase/id?id=",
        live: "https://api.blooket.com/api/firebase/live",
        verify: "https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyCustomToken?key=AIzaSyCA-cTOnX19f6LFnDVVsHXya3k6ByP_MnU",
        verifyAcc: "https://api.blooket.com/api/users/verify-session",
        history: "https://api.blooket.com/api/history",
        rewards: "https://api.blooket.com/api/users/add-rewards",
        isFavorited: "https://api.blooket.com/api/users/favorited?id=",
        favorite: "https://api.blooket.com/api/users/updatefavorites",
        games: "https://api.blooket.com/api/games",
        gameQuery: "https://api.blooket.com/api/games?gameId=",
        login: "https://api.blooket.com/api/users/login",
        homeWorkQuery: "https://api.blooket.com/api/homeworks?id=",
        homeworks: "https://api.blooket.com/api/users/homeworks",
        histories: "https://api.blooket.com/api/users/histories",
        blooks: "https://api.blooket.com/api/users/blooks",
        tokens: "https://api.blooket.com/api/users/tokens",
        users: "https://api.blooket.com/api/users",
        userQuery: "https://api.blooket.com/api/users?name=",
        unlock: "https://api.blooket.com/api/users/unlockblook",
        sell: "https://api.blooket.com/api/users/sellblook"
    },
};
